File: python/book-store/book_store.py
import collections
from itertools import combinations
from functools import lru_cache

PRICE = 800
DISCOUNTS = {
    1: 1.00,  # 0%
    2: 0.95,  # 5%
    3: 0.90,  # 10%
    4: 0.80,  # 20%
    5: 0.75,  # 25%
}
# Calculate cost per group size (integer cents)
GROUP_COSTS = {
    size: int(size * PRICE * discount) for size, discount in DISCOUNTS.items()
}
GROUP_COSTS[0] = 0 # Cost of an empty group is 0

# Use lru_cache for memoization. The state is defined by the counts of each book.
# We need a hashable representation of the counts, like a tuple.
@lru_cache(maxsize=None)
def calculate_min_cost(book_counts_tuple):
    """
    Recursively calculates the minimum cost for the given book counts.

    Args:
        book_counts_tuple: A tuple representing the counts of each book (e.g., (2, 1, 3, 0, 1)).
                           The order corresponds to book IDs 1 through 5.

    Returns:
        The minimum cost in cents.
    """
    if sum(book_counts_tuple) == 0:
        return 0

    min_total = float('inf')

    # Convert tuple back to a mutable dictionary or Counter for easier manipulation
    # Ensure keys 1-5 exist for consistent iteration later
    current_counts_dict = {i: book_counts_tuple[i-1] for i in range(1, 6)}

    # Find books currently available (count > 0)
    available_books = [book for book, count in current_counts_dict.items() if count > 0]
    num_distinct_available = len(available_books)

    # Iterate through possible group sizes to form (from 1 up to num_distinct_available)
    # Optimization: Start from largest possible group size? Might prune search space faster.
    # Let's stick to 1..N for now for correctness.
    for group_size in range(1, num_distinct_available + 1):
        # Iterate through all combinations of 'group_size' distinct books from the available ones
        for group_combination in combinations(available_books, group_size):
            # Create the next state of counts after forming this group
            next_counts_list = list(book_counts_tuple) # Mutable copy
            for book in group_combination:
                next_counts_list[book-1] -= 1 # Decrement count at the correct index

            next_counts_tuple = tuple(next_counts_list)

            # Recursive call
            cost = GROUP_COSTS[group_size] + calculate_min_cost(next_counts_tuple)
            min_total = min(min_total, cost)

    # If min_total remained infinity, something went wrong (shouldn't happen if input is valid)
    return min_total if min_total != float('inf') else 0


def total(basket):
    """
    Calculates the minimum cost for a basket of books applying group discounts.

    Args:
        basket: A list of integers representing book IDs (1 to 5).

    Returns:
        The minimum total cost in cents.
    """
    if not basket:
        return 0

    initial_counts = collections.Counter(basket)
    # Ensure counts for all 5 book types exist, even if 0, for consistent tuple representation
    initial_counts_tuple = tuple(initial_counts.get(i, 0) for i in range(1, 6))

    # Clear cache before each top-level call if running multiple tests in one process
    # This is important if the test runner reuses the same process.
    calculate_min_cost.cache_clear()

    return calculate_min_cost(initial_counts_tuple)

File: go/book-store/book_store.go
package bookstore


const bookPrice = 800 // Price in cents

// discountCosts maps the number of distinct books in a set to the total cost of that set in cents.
var discountCosts = map[int]int{
	1: 1 * bookPrice,                            // 0% discount
	2: 2 * bookPrice * 95 / 100,                 // 5% discount
	3: 3 * bookPrice * 90 / 100,                 // 10% discount
	4: 4 * bookPrice * 80 / 100,                 // 20% discount
	5: 5 * bookPrice * 75 / 100,                 // 25% discount
}

// Cost calculates the minimum cost for a basket of books considering discounts.
func Cost(books []int) int {
	if len(books) == 0 {
		return 0
	}

	// Count occurrences of each book
	bookCounts := make(map[int]int)
	for _, book := range books {
		bookCounts[book]++
	}

	// Form groups of distinct books greedily
	var groups []int
	for len(bookCounts) > 0 {
		currentGroupSize := 0
		var booksInGroup []int
		for bookID, count := range bookCounts {
			if count > 0 {
				currentGroupSize++
				booksInGroup = append(booksInGroup, bookID)
			}
		}

		if currentGroupSize == 0 {
			break // Should not happen if len(bookCounts) > 0 initially, but good practice
		}

		groups = append(groups, currentGroupSize)

		// Decrement counts for books included in this group
		for _, bookID := range booksInGroup {
			bookCounts[bookID]--
			if bookCounts[bookID] == 0 {
				delete(bookCounts, bookID) // Remove book type if count reaches zero
			}
		}
	}

	// Optimization: Prefer two groups of 4 over a group of 5 and a group of 3
	// Count 5s and 3s
	count5 := 0
	count3 := 0
	otherGroups := []int{}
	for _, size := range groups {
		switch size {
		case 5:
			count5++
		case 3:
			count3++
		default:
			otherGroups = append(otherGroups, size)
		}
	}

	// Perform replacements
	replacements := 0
	if count5 < count3 {
		replacements = count5
	} else {
		replacements = count3
	}


	optimizedGroups := otherGroups
	for i := 0; i < replacements; i++ {
		optimizedGroups = append(optimizedGroups, 4, 4) // Add two 4s for each (5, 3) pair
	}
	// Add remaining 5s and 3s
	for i := 0; i < count5-replacements; i++ {
		optimizedGroups = append(optimizedGroups, 5)
	}
	for i := 0; i < count3-replacements; i++ {
		optimizedGroups = append(optimizedGroups, 3)
	}


	// Calculate total cost from optimized groups
	totalCost := 0
	for _, size := range optimizedGroups {
		cost, exists := discountCosts[size]
		if exists {
			totalCost += cost
		}
		// Note: If size is 0 or > 5, it indicates an error in grouping logic,
		// but based on the algorithm, this shouldn't happen.
	}

	return totalCost
}
